builtin/credential/appgroup/user_id.go

```diff
@@ -0,0 +1,372 @@
+	}
+
+	// From the selector field supplied as the credential, detect the type of UserID
+	// supplied. UserID will be verified based on the type.
+	selectorType := ""
+	selectorValue := ""
+	switch {
+	case selector == "generic":
+		selectorType = "generic"
+		selectorValue = b.salt.SaltID(userID)
+	case strings.HasPrefix(selector, "app/") || strings.HasPrefix(selector, "group/"):
+		selectorFields := strings.SplitN(selector, "/", 2)
+		if len(selectorFields) != 2 {
+			return nil, fmt.Errorf("invalid selector; selector type and value could not be parsed")
+		}
+		selectorType = strings.TrimSpace(selectorFields[0])
+		selectorValue = strings.TrimSpace(selectorFields[1])
+		if selectorValue == "" {
+			return nil, fmt.Errorf("missing selector value")
+		}
+	default:
+		return nil, fmt.Errorf("unrecognized selector")
+	}
+
+	// Do the selector validation first. If this results in an error, the UserID
+	// entry should not be modified. Return the validation response if the UserID
+	// is found to be valid and if the UserID entry is updated properly.
+	validationResp, err := b.validateSelector(s, selectorType, selectorValue)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check if the user ID supplied is valid. If use limit was specified
+	// on the user ID, it will be decremented in this call.
+	valid, err := b.userIDEntryValid(s, selectorType, selectorValue, userID)
+	if err != nil {
+		return nil, err
+	}
+	if !valid {
+		return nil, fmt.Errorf("user ID not found under the %s selector type", selectorType)
+	}
+
+	return validationResp, nil
+}
+
+// Check if there exists an entry in the name of selectorValue for the selectorType supplied.
+// Prepares a response containing the combined set of policies, TTL and Wrapped values that are
+// applicable to the login.
+func (b *backend) validateSelector(s logical.Storage, selectorType, selectorValue string) (*validationResponse, error) {
+	resp := &validationResponse{
+		SelectorType:  selectorType,
+		SelectorValue: selectorValue,
+	}
+	switch selectorType {
+	case selectorTypeApp:
+		app, err := b.appEntry(s, selectorValue)
+		if err != nil {
+			return nil, err
+		}
+		if app == nil {
+			return nil, fmt.Errorf("app referred by the user ID does not exist")
+		}
+		resp.Policies = app.Policies
+		resp.TokenTTL = app.TokenTTL
+		resp.TokenMaxTTL = app.TokenMaxTTL
+		resp.Wrapped = app.Wrapped
+	case selectorTypeGroup:
+		group, err := b.groupEntry(s, selectorValue)
+		if err != nil {
+			return nil, err
+		}
+		if group == nil {
+			return nil, fmt.Errorf("group referred by the user ID does not exist")
+		}
+		groupPolicies, err := b.fetchPolicies(s, group.Apps)
+		if err != nil {
+			return nil, err
+		}
+		// Append the union of all the policies from all the apps on the group
+		resp.Policies = append(resp.Policies, groupPolicies...)
+
+		// Append the additional policies set on the group
+		resp.Policies = append(resp.Policies, group.AdditionalPolicies...)
+
+		resp.TokenTTL = group.TokenTTL
```

Comment:
I thought of this as well and share your opinion that it could lead to privilege escalation if operators are not cautious while creating groups. It can get difficult to reason about which options are finally being applied on the group, if we start to compare the options on the group with the options on the app.